Requirements.md

```diff
@@ -0,0 +1,101 @@
+# Requirements for GC
+
+This document identifies the key requirements for the managed memory proposal. Below we refer to the version of WASM with managed memory as being WASM-GC.
+
+## Goals
+
+1. Support for host-provided GC
+
+   Allow language implementers to use host-provided GC features as their primary means of supporting memory management. 
+   
+   On the whole, it would be preferable if language implementers could target WASM-GC as a natural target for their language. While it is difficult to be precise about this, an ideal scenario would be that language implementers could target WASM-GC as though it were just another back-end. In particular, targeting WASM-GC should try to avoid requiring changes to earlier stages of a compiler - such as the type checker or source-level transformations.
+   
+### Non-Goals
+1. Support for linear memory GC.
+
+   We do not attempt with this proposal to support alternate memory management strategies that rely on collecting linear memory.
+   
+## Criteria
+
+A criterion is intended to act as a metric of a 'soft value', such as 'desireability' or 'implementability'. Criteria are especially important when weighing the relative merits of features and in determining the importance that some scenario is satisfied.
+
+1. Prioriterize popular languages.
+   When deciding which features to provide it is important to place emphasis on features that are more likely to be used. We prioriterize support for the popular programming languages; while it is difficult to be precise about this, one measure is a public index such as the TIOBE index. The TIOBE ranking of the top 12 languages with managed memory (from the top 20 languages) is Java, Python, C#, Visual Basic, PHP, SQL, Swift, Go, Ruby, MATLAB, Perl, Scratch.
+   
+   Within these languages, there are languages which are (largely) statically typed -- such as Java and C# -- and languages that are dynamically typed -- such as Python. We prioriterize statically typed languages over dynamically typed language; because those languages are often more performance sensitive, and because support for statically typed languages can often be used for dynamically typed languages.
```

Comment:
The fundamental goal of prioritizing by language popularity is to maximize value for the WebAssembly ecosystem in the utilitarian sense. Language popularity is really just a proxy that lets us predict how much a particular language will be used with WebAssembly, but of course there are other factors as well, such as how interested the language maintainers are in targeting WebAssembly and how interested Web developers will be in using it. @fgmccabe, how would you feel about adding additional language after "We prioritize support for the popular programming languages" to be inclusive of additional languages that will have an outsized impact on the WebAssembly ecosystem? Possible additions might be:

 - "... and languages that appeal to Web developers"
 - "... and languages designed to interoperate with JavaScript"
 - "... and languages with strong interest in targeting WebAssembly"

Reply:
Elixir and Erlang are not even in the top 50 according to TIOBE 😢 , but [Lumen](https://github.com/lumen/lumen) (our Elixir/Erlang ahead-of-time compiler) is built on top of Rust, so if I can use it from RustI can just wrap that.

> how interested the language maintainers are in targeting WebAssembly

We've been members of the CG for over a year

Reply:
@KronicDeth can you elaborate on how you would like to be able to use the GC proposal from Rust? I think it is unlikely that LLVM will support the GC proposal at all, except perhaps in its assembly format.

Reply:
@tlively we would like to get away from the JS shim as much as possible. Right now we expect to still lean on rustwasm for the JS shim generation or something similar, but Lumen is directly compiling the Erlang itself using LLVM and then linking in the runtime, which is written in Rust.  If we have a DOM resource, we'd like to interact with them through this proposal and/or Interface Types if possible to eliminate the shimming.

Reply:
You may need to use a non-LLVM code generation backend if you aim to compile the entire language to the GC proposal. Would that be a deal breaker for you? If you only use the GC proposal for JS interop, it might be possible to use LLVM for your main codegen and link in some separately-generated bridge layer between the linear memory LLVM code and the GC interop code. It's not clear where that bridge code would come from or what it would look like, though, or whether using a GC proposal in that model would have any benefit over just using reference types.

Reply:
We're using LLVM so that we can build on top of MLIR and LLVM for the native compile target too.  We want to be able to target WASM, desktop/server, and embedded easily and LLVM gave us that.   I can't see switching the backend for GC as it is unlikely that generic WASM GC will be as useful as the GC that understands our type layout the way the one we wrote it.  It is more that we can interact correctly with stuff from Web APIs.

Reply:
> it is unlikely that generic WASM GC will be as useful as the GC that understands our type layout the way the one we wrote it. It is more that we can interact correctly with stuff from Web APIs.

If this is a common viewpoint among language implementors, the GC proposal will have failed. It would be helpful to understand in depth how the current proposal falls short of your needs and what it would take to fix that.

Reply:
@tlively I don't think it is quite that simple. Yes, a generic GC cannot ever cover all the functionality all specialized GCs that currently exist rely on, that is an impossible design exercise, so in that sense the Wasm GC will indeed fail. That said, there will be lots of new languages, languages without too strict performance requirements, language that needed to be reimplemented for Wasm anyway that will be very well served by the Wasm GC.

tl;dr: The Wasm ecosystem will be stronger with a generic Wasm GC, even if it doesn't serve everyone perfectly. But we should also take features for those that need to create their own GC in linear memory seriously.

Reply:
Erlang GCs way more often than I think a browser's GC is expecting most languages to GC.  In Erlang, each Erlang Process GCs only its own heap and we've already done tests where a single browser tab can support 64k Erlang processes.  A lot of times an Erlang process may never actually GC and instead only throw out it's whole heap when done.  (This is one of the reasons why Erlang calls it a "process" - it is isolated and you can just throw away the memory to free it when done like an OS process.)  Because the processes can send message to each other there are also HeapFragments for messages sitting in the timer system that can maybe be thought of as global, but all the terms in one HeapFragment are in each HeapFragment.  I'm expecting that the GC proposal is going to think in terms of GC for the entire thread/tab in a browser, so GCing one Erlang process will GC and lock the whole tab in some way.  If I'm wrong about this interpretation, let me know.

If LLVM is never going to support this GC proposal and If it is a trade-off between using LLVM to get all the code optimizations and using MLIR to get transparentish SIMD support, we're going to favor using LLVM than getting a GC, but having to lower from our IR to WASM directly.  We really want to avoid WASM being a vastly different target for the lowering.  That's one of these reasons we're pushing for stack switching because we do stack switching for x86_64 desktop/server.

Reply:
> That said, there will be lots of new languages, languages without too strict performance requirements, language that needed to be reimplemented for Wasm anyway that will be very well served by the Wasm GC.

My understanding is that we want the GC proposal to be usable by performance-sensitive languages with minimal changes to their existing compilers, but that should certainly be clarified in this requirements doc.



> If LLVM is never going to support this GC proposal and If it is a trade-off between using LLVM to get all the code optimizations and using MLIR to get transparentish SIMD support, we're going to favor using LLVM than getting a GC, but having to lower from our IR to WASM directly. We really want to avoid WASM being a vastly different target for the lowering.

We might be able to find a third option where we build an LLVM-like API on top of Binaryen. That would let you treat it very similarly to the LLVM backend and you would still get optimizations, but you would also be able to use GC and other proposals that don't fit well into LLVM.

But if Erlang is happy to use its own GC and it can't form cycles with JS (you said in a separate thread that Erlang doesn't have cycles), then what's still missing from the reference types proposal for JS interop? Is it the ability to pass Erlang objects out and use them from JS?

Reply:
> Is it the ability to pass Erlang objects out and use them from JS?

We don't currently have a story for transparent usage of Erlang types outside of the Lumen module.  For Rust `wasm_bindgen_test` we have a function that converts the Erlang types to sorta equivalents in JS, but Erlang has tuples and JS doesn't and JS has no concept of a PID (Process IDentifier, used to interact with a process from another process), so it's not transparent and not a safe thing where we know we can accept it back in from JS.  We brainstormed about doing something tricky with Proxies in JS to make it Just Work™️ , but haven't prototyped it.

Reply:
@KronicDeth 

> Erlang GCs way more often than I think a browser's GC is expecting most languages to GC. In Erlang, each Erlang Process GCs only its own heap and we've already done tests where a single browser tab can support 64k Erlang processes. 

I think that this does suggest a requirement that I had in the back of my mind: support for multiple GC memories; to follow from the core WASM multiple memory proposal.
However, 64K different memories is significantly more than the authors of that proposal were thinking, I believe!

Reply:
@tlively 

> My understanding is that we want the GC proposal to be usable by performance-sensitive languages with minimal changes to their existing compiler

There seems to be a significant tension here. On the one hand, a highly performing compiler 'knows' how to express desired optimizations and it behooves the WASM to get out of the way and let it rip. OTOH, the WASM engine exists in a context that may contradict the smart compiler's assumptions. 

I believe the answer to this is @titzer's idea of layering. Let the compiler lower their language to a point where the engine can straightforwardly take over.

At the risk of incurring @sabine's wrath, this probably means that tagged pointers are not the way forward - because the risk of clashing with the WASM engine's existing context is too high.

Reply:
First off, I'm not going to be angry if tagged pointers don't exist in WASM GC.

OCaml needs a reasonably efficient mechanism that allows storing either a reference or a scalar in the same field of a heap struct, and checking at runtime, which of the two a value is.

If we knew for certain that linear memory GC will be supported eventually, we could focus our efforts on that, instead of trying to push for a heap model that works for us.